devito/operator.py

```diff
@@ -114,63 +119,60 @@ def __init__(self, expressions, **kwargs):
-        """
-        default_args = ArgumentMap()
-        for p in self.input:
-            if p.name not in arguments:
-                default_args.update(p.argument_defaults())
-        for p in self.dimensions:
-            if p.name not in arguments and p.is_Sub:
-                default_args.update(p.argument_defaults(default_args))
-        return {k: default_args.reduce(k) for k in default_args if k not in arguments}
-
-    def arguments(self, **kwargs):
+    def prepare_arguments(self, **kwargs):
         """
         Process runtime arguments passed to ``.apply()` and derive
         default values for any remaining arguments.
         """
-        arguments = {}
-        # First, we insert user-provided override
-        for p in self.input + self.dimensions:
-            arguments.update(p.argument_values(**kwargs))
+        # Handle data-carriers (first overrides, then defaults)
+        args = ReducerMap()
+        args.update([p._arg_values(**kwargs) for p in self.input if p.name in kwargs])
+        args.update([p._arg_defaults() for p in self.input if p.name not in args])
+        args = args.reduce_all()
 
-        # Second, derive all remaining default values from parameters
-        arguments.update(self._argument_defaults(arguments))
+        # Handle dimensions (first adjust data-carriers-induced defaults, then overrides)
+        for p in self.dimensions:
+            args.update(p._arg_infers(args, self._dspace[p], **kwargs))
+        for p in self.dimensions:
+            args.update(p._arg_values(**kwargs))
+
+        # Sanity check
+        for p in self.input:
+            p._arg_check(args, self._dspace[p])
 
         # Derive additional values for DLE arguments
         # TODO: This is not pretty, but it works for now. Ideally, the
         # DLE arguments would be massaged into the IET so as to comply
         # with the rest of the argument derivation procedure.
-        for arg in self.dle_arguments:
+        for arg in self.dle_args:
             dim = arg.argument
-            osize = arguments[arg.original_dim.symbolic_size.name]
+            osize = args[arg.original_dim.symbolic_size.name]
             if dim.symbolic_size in self.parameters:
                 if arg.value is None:
-                    arguments[dim.symbolic_size.name] = osize
+                    args[dim.symbolic_size.name] = osize
                 elif isinstance(arg.value, int):
-                    arguments[dim.symbolic_size.name] = arg.value
+                    args[dim.symbolic_size.name] = arg.value
                 else:
-                    arguments[dim.symbolic_size.name] = arg.value(osize)
+                    args[dim.symbolic_size.name] = arg.value(osize)
 
         # Add in the profiler argument
-        arguments[self.profiler.name] = self.profiler.new()
+        args[self.profiler.name] = self.profiler.new()
+
+        # Add in any backend-specific argument
+        args.update(kwargs.get('backend', {}))
```

Comment:
You should always be using `kwargs.pop` to access `kwargs` and then ensure that it is empty in the end.